### ISSUE TITLE
LevenbergMarquardt keeps a rejected step when line search fails

### ISSUE THREAD

**Issue by kovacoj** [kovacoj1/optimizers#27]
## Summary
`LevenbergMarquardt.step()` applies an update before the line search loop, and if none of the retried damped updates reduce the loss, it exits with the last attempted update still applied.

## Why this is a correctness bug
The line search is supposed to keep an update only when it improves the residual loss. In the current implementation, when every candidate step is rejected, the optimizer still leaves the final rejected step in the model parameters.

That means `step()` can increase the loss even though its internal acceptance check never succeeded.

## Current code
- `src/optimizers/LevenbergMarquardt.py:77` applies the initial update immediately
- `src/optimizers/LevenbergMarquardt.py:81-97` retries larger `mu` values inside the loop
- `src/optimizers/LevenbergMarquardt.py:99-105` only adjusts `mu` after success, but never restores the original parameters if the loop exhausts without success

## Reproduction
This reproduces on the current `main`:

```python
import torch
from optimizers import LevenbergMarquardt

x = torch.nn.Parameter(torch.tensor([-2.9], dtype=torch.float64))
opt = LevenbergMarquardt([x], mu=1e-6, m_max=1)

def closure():
    return torch.sin(5 * x).view(-1)

before = (closure() @ closure()).item()
opt.step(closure)
after = (closure() @ closure()).item()
print(before, after, x.item())
```

Observed result:
- before: `0.8740287648445001`
- after: `0.9792091631715039`
- final `x`: `-3.4268129884259255`

## Expected behavior
If no trial step decreases the loss after the configured `m_max` retries, the optimizer should leave the parameters at the original point for that iteration.
